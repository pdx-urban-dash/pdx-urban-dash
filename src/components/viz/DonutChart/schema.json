{
  "$schema": "https://vega.github.io/schema/vega/v5.json",
  "width": 500,
  "height": 200,
  "padding": 5,

  "data": [
    {
      "name": "table",
      "$dataValues": "",
      "transform": [
        {
          "type": "pie",
          "field": "field",
          "startAngle": 0,
          "endAngle": 6.29,
          "sort": false
        }
      ]
    }
  ],

  "config": {
    "axis": {
      "$labelColor": "",
      "$labelFontSize": 0
    }
  },

  "scales": [
    {
      "name": "color",
      "type": "ordinal",
      "domain": {"data": "table", "field": "id"},
      "$colorRange": []
    }
  ],

  "marks": [
    {
      "type": "arc",
      "from": {"data": "table"},
      "encode": {
        "enter": {
          "fill": {"scale": "color", "field": "id"},
          "x": {"signal": "width / 2"},
          "y": {"signal": "height / 2"}
        },
        "update": {
          "startAngle": {"field": "startAngle"},
          "endAngle": {"field": "endAngle"},
          "innerRadius": {"signal": "width / 8"},
          "outerRadius": {"signal": "width / 4"}
        }
      }
    }
  ]
}