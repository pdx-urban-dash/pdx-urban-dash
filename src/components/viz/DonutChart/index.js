import React from 'react';
import PropTypes from 'prop-types';
import { createClassFromSpec } from 'react-vega';

import { injectPropsIntoSchema } from '../../../utils/vegaUtils';
import { colors, sizes } from '../../../constants';
import donutChartSchema from './schema';
import './styles.scss';

import Legend from '../shared/Legend';
import ChartWrapper from '../shared/ChartWrapper';
import { getDefaultColor } from '../../../utils/vizUtils';

const buildData = data => data.map((elem, idx) => ({
  id: idx,
  field: elem.value,
  c: idx,
}));

export const getColors = data => data.map((elem, idx) => (elem.color || getDefaultColor(idx)));

const DonutChart = (
  {
    data,
    title,
    className,
    colorScheme,
  },
) => {
  const mColors = getColors(data);
  const Graph = createClassFromSpec(
    injectPropsIntoSchema(
      {
        $dataValues: {
          name: 'values',
          value: buildData(data),
        },
        $labelColor: {
          name: 'labelColor',
          value: colorScheme === 'light' ? colors.fontDark : colors.fontLight,
        },
        $labelFontSize: {
          name: 'labelFontSize',
          value: parseInt(sizes.fontSM, 10),
        },
        $colorRange: {
          name: 'range',
          value: mColors,
        },
      },
      donutChartSchema,
    ),
  );
  return (
    <ChartWrapper title={title} className={className}>
      <Graph />
      <Legend
        series={data.map((elem, idx) => ({
          title: elem.name,
          description: elem.description,
          color: mColors[idx],
        }))}
      />
    </ChartWrapper>
  );
};

DonutChart.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
    color: PropTypes.string,
    value: PropTypes.number.isRequired,
  })).isRequired,
  title: PropTypes.string,
  className: PropTypes.string,
  colorScheme: PropTypes.oneOf(['light', 'dark']),
};

DonutChart.defaultProps = {
  title: 'default title',
  className: '',
  colorScheme: 'light',
};

export default DonutChart;
