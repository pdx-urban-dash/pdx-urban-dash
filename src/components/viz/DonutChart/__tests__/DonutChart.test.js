import React from 'react';
import { shallow } from 'enzyme';
import DonutChart from '../index';

describe('<DonutChart />', () => {
  test('snapshot', () => {
    const wrapper = shallow(
      <DonutChart
        title="title"
        className=""
        colorScheme="light"
        data={[
          {
            name: 'first slice',
            description: 'first description',
            value: 10,
          },
          {
            name: 'second slice',
            description: 'second description',
            value: 10,
          },
          {
            name: 'third slice',
            description: 'third description',
            value: 10,
          },
          {
            name: 'fourth slice',
            description: 'fourth description',
            value: 10,
          },
        ]}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
