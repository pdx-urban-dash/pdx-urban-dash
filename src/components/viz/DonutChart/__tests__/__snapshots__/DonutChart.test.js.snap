// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DonutChart /> snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DonutChart
    className=""
    colorScheme="light"
    data={
      Array [
        Object {
          "description": "first description",
          "name": "first slice",
          "value": 10,
        },
        Object {
          "description": "second description",
          "name": "second slice",
          "value": 10,
        },
        Object {
          "description": "third description",
          "name": "third slice",
          "value": 10,
        },
        Object {
          "description": "fourth description",
          "name": "fourth slice",
          "value": 10,
        },
      ]
    }
    title="title"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <a />,
        <Legend
          series={
            Array [
              Object {
                "color": "#82204A",
                "description": "first description",
                "title": "first slice",
              },
              Object {
                "color": "#558C8C",
                "description": "second description",
                "title": "second slice",
              },
              Object {
                "color": "#E8DB7D",
                "description": "third description",
                "title": "third slice",
              },
              Object {
                "color": "#231123",
                "description": "fourth description",
                "title": "fourth slice",
              },
            ]
          }
        />,
      ],
      "className": "",
      "title": "title",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "series": Array [
            Object {
              "color": "#82204A",
              "description": "first description",
              "title": "first slice",
            },
            Object {
              "color": "#558C8C",
              "description": "second description",
              "title": "second slice",
            },
            Object {
              "color": "#E8DB7D",
              "description": "third description",
              "title": "third slice",
            },
            Object {
              "color": "#231123",
              "description": "fourth description",
              "title": "fourth slice",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <a />,
          <Legend
            series={
              Array [
                Object {
                  "color": "#82204A",
                  "description": "first description",
                  "title": "first slice",
                },
                Object {
                  "color": "#558C8C",
                  "description": "second description",
                  "title": "second slice",
                },
                Object {
                  "color": "#E8DB7D",
                  "description": "third description",
                  "title": "third slice",
                },
                Object {
                  "color": "#231123",
                  "description": "fourth description",
                  "title": "fourth slice",
                },
              ]
            }
          />,
        ],
        "className": "",
        "title": "title",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "series": Array [
              Object {
                "color": "#82204A",
                "description": "first description",
                "title": "first slice",
              },
              Object {
                "color": "#558C8C",
                "description": "second description",
                "title": "second slice",
              },
              Object {
                "color": "#E8DB7D",
                "description": "third description",
                "title": "third slice",
              },
              Object {
                "color": "#231123",
                "description": "fourth description",
                "title": "fourth slice",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
